### PR TITLE
ios14 location permissions

### LINKED ISSUE #29: [ios] Location Permissions in iOS 14
Cherry pick from:
https://github.com/mapbox/mapbox-gl-native-ios/pull/376
https://github.com/mapbox/mapbox-gl-native-ios/pull/381
https://github.com/mapbox/mapbox-gl-native-ios/pull/403
https://github.com/mapbox/mapbox-gl-native-ios/pull/392
https://github.com/mapbox/mapbox-gl-native-ios/pull/474
https://github.com/mapbox/mapbox-gl-native-ios/pull/361

### PR DESCRIPTION
**Pull request by petr-pokorny-1**
resolves #29

### CODE REVIEW

File: platform/ios/platform/darwin/src/MGLLocationManager.m [hunk 1]
```diff
@@ -38,7 +38,16 @@ - (CLLocationAccuracy)desiredAccuracy {
...
 
 - (CLAuthorizationStatus)authorizationStatus {
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 140000
+    if (@available(iOS 14.0, *)) {
+        return self.locationManager.authorizationStatus;
+    } else {
+        return kCLAuthorizationStatusNotDetermined;
+    }
+#else
+    return [CLLocationManager authorizationStatus];
```
Comment (roblabs) [inline platform/ios/platform/darwin/src/MGLLocationManager.m:48]: @petr-pokorny-1  — there are two `return`s here: 

`return [CLLocationManager authorizationStatus];`

File: platform/ios/platform/ios/app/Info.plist [hunk 2]
```diff
@@ -26,12 +26,28 @@
 	<true/>
 	<key>NSHumanReadableCopyright</key>
 	<string>© 2014–2019 Mapbox</string>
+	<key>NSLocationAlwaysAndWhenInUseUsageDescription</key>
+	<string>The map will display your location. If you choose Always, the map may also use your location when it isn’t visible in order to improve OpenStreetMap and Mapbox products.</string>
```
Comment (roblabs) [inline platform/ios/platform/ios/app/Info.plist:30]: @petr-pokorny-1  - Potential change:   

* Remove the section ~~ and Mapbox products~~. 
* I don't even think changing to MapLibre would make sense

File: platform/ios/platform/ios/app/MBXViewController.m [hunk 3]
```diff
@@ -2333,6 +2338,33 @@ - (void)mapViewDidFinishRenderingFrame:(MGLMapView *)mapView fullyRendered:(BOOL
...
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 140000
+        if (manager.authorizationStatus == kCLAuthorizationStatusDenied || manager.accuracyAuthorization == CLAccuracyAuthorizationReducedAccuracy) {
+            [self alertAccuracyChanges];
+        }
+#endif
+    }
+}
+
+- (void)alertAccuracyChanges {
+    UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Mapbox GL works best with your precise location."
```
Comment (roblabs) [inline platform/ios/platform/ios/app/MBXViewController.m:2352]: @petr-pokorny-1  — potential fix:

@"MapLibre GL works best with your precise location."

**Review by roblabs — CHANGES_REQUESTED** [review]

**Comment by roblabs** [comment]
1.  It is **_really_** nice to have [automatic checks](https://github.com/maplibre/maplibre-gl-native/pull/31/checks).  @petr-pokorny-1 — great job on setting up the GitHub Actions for this repo.  I did some review there (but not sure how to look for issues)
2. Tested with the schemes:
* CI — warnings are related to GLKit
* iosapp — warnings are related to GLKit
* bench - failed, but I'm not sure if it failed before (seems likely)

---

_Desktop Xcode scheme testing summary_
image

File: platform/ios/platform/darwin/src/MGLLocationManager.m (hunk 1 above)
Comment (petr-pokorny-1) [reply platform/ios/platform/darwin/src/MGLLocationManager.m:48]: Thanks, good catch. Fixed

File: platform/ios/platform/ios/app/MBXViewController.m (hunk 3 above)
Comment (petr-pokorny-1) [reply platform/ios/platform/ios/app/MBXViewController.m]: text updated

File: platform/ios/platform/ios/app/Info.plist (hunk 2 above)
Comment (petr-pokorny-1) [reply platform/ios/platform/ios/app/Info.plist]: text updated

**Comment by petr-pokorny-1** [comment]
> 
>     * bench - failed, but I'm not sure if it failed before (seems likely)
> 

I was able to fix benchmark app and launch it on my iphone.

Screen Shot 2021-02-23 at 1 49 30 PM

**Comment by roblabs** [comment]
Bench mark results on x86_64 Simulator.  Thanks for fixing

Benchmark completed.
Result:

| Location | FPS |
| :---  | :--- |
| paris      | 11.5 fps |
| paris2     | 16.3 fps |
| alps       |  6.2 fps |
| us east    | 16.9 fps |
| greater la | 11.9 fps |
| sf         | 19.1 fps |
| oakland    | 13.3 fps |
| germany    |  7.6 fps |

Total FPS: 102.9
Average FPS: 12.9

**Pull request merged**